### PR TITLE
CLOUDP-148833: Support multiple atlas-operator versions

### PR DESCRIPTION
**Pull request by igor-karpukhin**
## Proposed changes

Added an ability to generate kubernetes resources for multiple versions of Atlas operator. Introduced a new flag for the `atlas kubernetes config generate` command: `--operatorVersion=...`

_Jira ticket:_ [CLOUDP-148833](https://jira.mongodb.org/browse/CLOUDP-148833)

Closes #[issue number]

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation in document requirements section listed in [CONTRIBUTING.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/CONTRIBUTING.md) (if appropriate)
- [ ] I have addressed the @mongodb/docs-cloud-team comments (if appropriate)
- [ ] I have updated [test/README.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/test/README.md) (if an e2e test has been added)
- [x] I have run `make fmt` and formatted my code

### CODE REVIEW

File: internal/kubernetes/operator/deployment/deployment.go
Comment (gssbzn) [inline internal/kubernetes/operator/deployment/deployment.go:22]:
```suggestion
	"github.com/mongodb/mongodb-atlas-cli/internal/kubernetes/operator/features"
```

File: internal/kubernetes/operator/features/crds.go
Comment (gssbzn) [inline internal/kubernetes/operator/features/crds.go:139]: `IsFeatureExist` doesn't read right, consider `IsValidFeature` or `FeatureExists`

File: internal/kubernetes/operator/deployment/deployment_test.go
Comment (gssbzn) [inline internal/kubernetes/operator/deployment/deployment_test.go:27]:
```suggestion
	"github.com/mongodb/mongodb-atlas-cli/internal/kubernetes/operator/features"
```

File: internal/kubernetes/operator/features/crds.go
Comment (gssbzn) [inline internal/kubernetes/operator/features/crds.go:64]: getters are non ideomatic
```suggestion
func SupportedVersions() []string {
```

File: internal/cli/atlas/kubernetes/config/generate.go
Comment (gssbzn) [inline internal/cli/atlas/kubernetes/config/generate.go:142]: looking at how `features.GetSupportedVersions()` is implemented I have some mild concerns always defaulting to the newest CRD may be risky not familiar enough with the CRD ecosystem to have a strong opinion but in the CLI we aim for the "least surprise" approach, meaning behavior of a command should not change between versions and the current approach lands in the "it'll change at random point" category

File: internal/kubernetes/operator/features/crds.go
Comment (gssbzn) [inline internal/kubernetes/operator/features/crds.go:99]: cloud/should we wrap this in the store layer?

File: internal/kubernetes/operator/features/mock.go
Comment (gssbzn) [inline internal/kubernetes/operator/features/mock.go:7]: @Sugar-pack recently moved all the operator mocks to autogenerated mocks, I think the team needs to pick a single approach

File: internal/kubernetes/operator/project/project_test.go
Comment (gssbzn) [inline internal/kubernetes/operator/project/project_test.go:26]:
```suggestion
	"github.com/mongodb/mongodb-atlas-cli/internal/kubernetes/operator/features"
```

**Review by gssbzn — CHANGES_REQUESTED** [review]
Left some comments, please also make sure to run `make addcopy` and finally this probably needs a review from the Operator team as I'm seeing some clashing patterns now

**Comment by igor-karpukhin** [comment]
Hi @gssbzn. Thanks for the review. Let's hold this PR for a bit until another PR with autogenerated mocks for atlas-operator resources is merged. I'll rebase this one on top of it and then I'll ask for a review one more time.

File: docs/atlascli/command/atlas-kubernetes-config-generate.txt
Comment (erabil-mdb) [inline docs/atlascli/command/atlas-kubernetes-config-generate.txt:113]:
```suggestion
   # Export resources for a specific version of Atlas Kubernetes Operator:
```

**Review by erabil-mdb — APPROVED** [review]
LGTM with a few docs suggestions!

File: internal/cli/atlas/kubernetes/config/generate.go
Comment (igor-karpukhin) [reply internal/cli/atlas/kubernetes/config/generate.go:142]: I understand your concern. But I'd expect a new version of the CLI to generate an output for a newer version of the operator. One possible solution I can think of is that I can make this flag `required` without a default version. That would force one to explicitly set this flag.

File: internal/cli/atlas/kubernetes/config/generate.go
Comment (igor-karpukhin) [reply internal/cli/atlas/kubernetes/config/generate.go]: Fixed

File: internal/cli/atlas/kubernetes/config/generate.go
Comment (dan-mckean) [reply internal/cli/atlas/kubernetes/config/generate.go:142]: We chatted this over. I'd like to go ahead with it as is. The worst case is that the Operator return to the user that they're trying to import something newer than the operator itself. We may iterate on this later to try and read from Kubernetes the actual version of the operator in use.

File: internal/kubernetes/operator/features/crds.go
Comment (igor-karpukhin) [reply internal/kubernetes/operator/features/crds.go]: Thanks! Fixed

File: internal/kubernetes/operator/features/crds.go
Comment (igor-karpukhin) [reply internal/kubernetes/operator/features/crds.go]: Fixed

File: internal/kubernetes/operator/features/mock.go
Comment (igor-karpukhin) [reply internal/kubernetes/operator/features/mock.go]: Fixed

File: internal/usage/usage.go
Comment (gssbzn) [inline internal/usage/usage.go:388]: This only became obvious because of the default value, can you make sure the helps end in a full stop (.)
```suggestion
	OperatorVersion                           = "Version of Atlas Kubernetes Operator to generate resources for."
```

File: internal/kubernetes/operator/features/crds.go
Comment (igor-karpukhin) [reply internal/kubernetes/operator/features/crds.go:113]: I thought about that, but the store seems be working with the Atlas API. I can do that, but I'm not sure if that is needed.

File: internal/usage/usage.go
Comment (igor-karpukhin) [reply internal/usage/usage.go]: Fixed, thanks!

File: internal/kubernetes/operator/features/crds.go
Comment (gssbzn) [reply internal/kubernetes/operator/features/crds.go:113]: When I designed the CLI I'd did with 3 core layers, the config layer, one the settings, the store layer, or how to access data and the CLI layer, the thing that flies them (and other parts) together.
You can see store as a transparent way to access data (DAO?) In the CLI no matter where the data is, proof of this model, the store is able to handle atlas or ops manager in a transparent way for consumers. This approach let's us have some decoupling between the biz logic and data and why I'm asking here if long term you can benefit of the decoupling both

**Review by gssbzn — APPROVED** [review]
Lgtm

File: internal/kubernetes/operator/features/crds.go
Comment (igor-karpukhin) [reply internal/kubernetes/operator/features/crds.go:113]: Yes, I noticed this approach, I just didn't want to extend it with this small feature, as it has nothing to do with the Atlas API. But okay, let me add it to the `Store`.

Comment (igor-karpukhin) [reply internal/kubernetes/operator/features/crds.go]: Fixed by moving CRDs download logic into a separate provider

File: internal/kubernetes/operator/dbusers/dbusers_test.go
Comment (gssbzn) [inline internal/kubernetes/operator/dbusers/dbusers_test.go:230]: any reason to use panic instead of t.Fatal?

**Review by gssbzn — COMMENTED** [review]
couple of small things, this still needs a review from the operator side as I'm not going deep on that part

File: internal/cli/atlas/kubernetes/config/generate.go
Comment (igor-karpukhin) [reply internal/cli/atlas/kubernetes/config/generate.go]: Fixed

File: internal/kubernetes/operator/dbusers/dbusers_test.go
Comment (igor-karpukhin) [reply internal/kubernetes/operator/dbusers/dbusers_test.go]: Fixed

**Review by gssbzn — APPROVED** [review]
LGTM

**Comment by igor-karpukhin** [comment]
@fabritsius please take a look as well

**Review by fabritsius — APPROVED** [review]
Looks Great, Massive Work Here 👍

**Pull request merged**
